networkx/algorithms/bipartite/envy_free_matching.py

```diff
@@ -0,0 +1,375 @@
+    finding envy free bipartite matching, this is proved in the article.
+
+    Two sets are calculated with the following recursive definition:
+
+    X_0 = nodes in G unmatched by M.
+    for i>=1:
+        Latex:
+            Y_i = :math:`N_{G\M}(X_{i-1})\\`
+            X_i = :math:`N_M(Y_i)`
+        English:
+            Y_i -> neighbors of X_{i-1} that were not matched by M
+                   group-subtracting all the previous Y_k's we've seen
+            X_i -> neighbors of Y_i that were matched by M
+
+    The iterative pattern stops when either Y_k or X_k are an empty set for a
+    certain k, forever staying an empty set. From k onward the alternating sequence
+    is proven to always stop in the article.
+
+    Two tuples containing sets of nodes are returned - tuple(X_subsets), tuple(Y_subsets).
+
+    >>> G = nx.complete_bipartite_graph(3, 3)
+    >>> M = nx.bipartite.hopcroft_karp_matching(G)
+    >>> _M_alternating_sequence(G, M)
+    ((), ())
+
+    >>> G = nx.Graph([(0, 3), (0, 4), (1, 4), (2, 4)])
+    >>> M = {0: 3, 3: 0, 1: 4, 4: 1}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}), ({4},))
+
+    >>> G = nx.Graph([(0, 3), (1, 3), (1, 4), (2, 4)])
+    >>> M = {0: 3, 3: 0, 4: 1, 1: 4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0, 3), (1, 3), (1, 4), (2, 4)])
+    >>> M = {0: 3, 3: 0, 4: 1, 1: 4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph(
+    ...     [(0, 6), (1, 6), (1, 7), (2, 6), (2, 8), (3, 9), (3, 6), (4, 8), (4, 7), (5, 9)]
+    ... )
+    >>> M = {0: 6, 6: 0, 1: 7, 7: 1, 2: 8, 8: 2, 3: 9, 9: 3}
+    >>> _M_alternating_sequence(G, M)
+    (({4, 5}, {1, 2, 3}, {0}), ({8, 9, 7}, {6}))
+
+    """
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_0 = set(X) - set(M.keys())
+    if len(X_0) == 0:
+        return (), ()
+
+    X_subsets = [X_0]
+    Y_subsets = []
+
+    Y_subgroups_accumulate = set()
+
+    while len(X_subsets) == 1 or (len(X_subsets[-1]) != 0 and len(Y_subsets[-1]) != 0):
+        Y_current = {
+            nbr
+            for n in X_subsets[-1]
+            for nbr in G[n]
+            if nbr not in Y_subgroups_accumulate
+            if M.get(n) != nbr
+            if M.get(nbr) != n
+        }
+        if len(Y_current) == 0:
+            break
+        X_current = {
+            nbr
+            for n in Y_current
+            for nbr in G[n]
+            if M.get(n) == nbr and M.get(nbr) == n
+        }
+        if len(X_current) == 0:
+            break
+
+        Y_subsets.append(Y_current)
+        X_subsets.append(X_current)
+
+        Y_subgroups_accumulate.update(Y_current)
+
+    return tuple(X_subsets), tuple(Y_subsets)
+
+
+def envy_free_matching_partition(G, *, M=None, top_nodes=None):
+    r"""Return a unique EFM (Envy Free Matching) partition of bipartite graph `G`.
```

Comment:
"""
Returns the envy-free-matching partition (EFM partition) of the given bipartite graph `G`.

Given a bipartite graph G = (X+Y, E), an EFM partition is a partition of the nodes into two disjoint parts: XS+YS and XL+YL, satisfying the following conditions:

(a) There are no edges between XS and YL ;
(b) The subgraph G[XL,YL] is X-saturated; this means that it contains a matching in which all vertices of XL are matched.
(c) The subgraph G[XS,YS] is Y-path-saturated; this means that it contains a structure similar to an odd path, which guarantees that the only EFM it contains is the empty matching.

Intuitively, G[XL,YL] is the "good" part of G, that contains the largest possible EFM; G[XS,YS] is the "bad" part of G, which contains only an empty EFM. 

In the paper [1], it is proved that there is a *unique* partition satisfying properties (a), (b), (c). Moreover, this this unique partition has the following additional properties:
(d) Every maximum matching in G[XL,YL] is an EFM in G.
(e) Every EFM in G is contained in G[XL,YL].